h2/src/main/org/h2/engine/Database.java

```diff
@@ -335,6 +338,14 @@ public int getLockTimeout() {
         return setting == null ? Constants.INITIAL_LOCK_TIMEOUT : setting.getIntValue();
     }
 
+    public RowFactory getRowFactory() {
+        return rowFactory;
+    }
+
+    public void setRowFactory(RowFactory rowFactory) {
+        this.rowFactory = rowFactory;
+    }
```

Comment:
Personally I don't think that the whole database should have that setting. H2 has two own table engines that can be used together (PS with spatial index), and few people plug additional table engines for some tables. Custom implementation set by someone may break some of engines. If MV engine needs such configuration it can be moved to it.

Also some implementations may want to use specialized rows for each table, so the common code should ask a table to construct a row for it as current H2 tries to do.

Reply:
Of course, it isn't critical, because custom engines may override all related methods from `Table`. But `Table.setColumns()` is also affected by that change.

Maybe only `MVTable` should have such logic instead of the base `Table` class.

Reply:
We don't have a notion of engine-specific settings yet, and Row produced by RowFactory engine-agnostic and is used by all (2) engines. I think it's an abstraction that is orthogonal to Engine, at least for now. Factory could behave differently for different tables, at least it has opportunity to do so. For that Table has to remember Engine which produced it. Behavior can be customized even at Index level (BaseIndex constructor has a call to createFactory()).
So maybe Table should have a reference to the Engine, that produced and it can be passed to RowFactory.createRowFactory() ? Any custom Engine will produce a Table which depends on a common notion of Row, so why RowFactory should be engine-specific?

Reply:
From my point of view normal tables (`RegularTable` subclasses) can remember own engine and use its methods, if you like that approach I won't complain.

Different virtual tables aren't stored anywhere and they don't belong to any engine.